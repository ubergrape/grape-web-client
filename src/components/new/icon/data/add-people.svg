<svg width="16" height="12" viewBox="0 0 16 12" xmlns="http://www.w3.org/2000/svg">
<path d="M10.1818 6.00002C11.7891 6.00002 13.0909 4.74702 13.0909 3.20002C13.0909 1.65302 11.7891 0.400024 10.1818 0.400024C8.57456 0.400024 7.27274 1.65302 7.27274 3.20002C7.27274 4.74702 8.57456 6.00002 10.1818 6.00002ZM3.63639 4.60002V2.50002H2.18184V4.60002H3.05176e-05V6.00002H2.18184V8.10003H3.63639V6.00002H5.8182V4.60002H3.63639ZM10.1818 7.40002C8.24002 7.40002 4.36366 8.33803 4.36366 10.2V11.6H16V10.2C16 8.33803 12.1236 7.40002 10.1818 7.40002Z"/>
</svg>
