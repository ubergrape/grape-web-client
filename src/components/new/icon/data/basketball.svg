<svg width="32" height="32" viewBox="0 0 32 32" xmlns="http://www.w3.org/2000/svg"><g fill-rule="evenodd"><path d="M15.833 25.267a9.433 9.433 0 1 0 0-18.867 9.433 9.433 0 0 0 0 18.867zm0-1.2a8.233 8.233 0 1 1 0-16.467 8.233 8.233 0 0 1 0 16.467z"/><path d="M11.4 7.833c0 3.258 3.966 5.767 8.767 5.767 1.356 0 2.67-.2 3.86-.58l-.366-1.144a11.5 11.5 0 0 1-3.49.524c-4.22 0-7.56-2.117-7.56-4.567h-1.2zM20.443 23.5c0-3.257-3.966-5.767-8.766-5.767-1.356 0-2.67.2-3.86.58l.366 1.144a11.54 11.54 0 0 1 3.494-.524c4.22 0 7.566 2.117 7.566 4.567h1.2zm3.992-5.502c-2.532-1.223-5.425-2.344-8.49-3.28-2.92-.894-5.804-1.56-8.48-1.97l-.18 1.186c2.616.4 5.444 1.055 8.31 1.93 3.007.92 5.843 2.02 8.32 3.214l.52-1.08z"/><path d="M19.245 7.466c-2.993 2.302-5.088 6.04-5.687 10.3-.338 2.403-.175 4.79.453 6.98l1.16-.33c-.58-2.03-.73-4.246-.42-6.484.56-3.954 2.49-7.41 5.23-9.514l-.73-.952z"/></g></svg>
