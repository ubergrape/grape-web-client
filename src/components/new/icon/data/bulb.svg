<svg width="32" height="32" viewBox="0 0 32 32" xmlns="http://www.w3.org/2000/svg"><g fill-rule="evenodd"><path d="M16.188 7.5A6.187 6.187 0 0 0 10 13.688c0 2.267 2.075 4.67 2.83 6.78 1.127 3.148 1.002 5.032 3.357 5.032 2.39 0 2.23-1.875 3.358-5.02.758-2.114 2.83-4.543 2.83-6.793A6.188 6.188 0 0 0 16.187 7.5zm1.457 15.287l-2.788.348a23.85 23.85 0 0 1-.337-1.065c0-.006-.004-.012-.005-.018l3.48-.434c-.05.166-.1.34-.148.496-.073.253-.14.472-.202.672zm-3.293-1.276a33.75 33.75 0 0 0-.346-1.07h4.368c-.07.2-.138.41-.198.6l-3.824.48zm1.835 2.87c-.57 0-.83-.062-1.123-.7l2.384-.3c-.345.93-.597 1.004-1.26 1.004zm2.628-5.06h-5.25a26.847 26.847 0 0 0-.948-1.81c-.734-1.312-1.492-2.67-1.492-3.813a5.068 5.068 0 0 1 5.063-5.063 5.068 5.068 0 0 1 5.062 5.063c0 1.133-.76 2.5-1.494 3.82-.33.592-.663 1.2-.94 1.802z"/><path d="M16 9.9c-1.59 0-2.952 1.22-3.565 2.95l1.13.4c.46-1.29 1.417-2.15 2.435-2.15V9.9z"/></g></svg>
