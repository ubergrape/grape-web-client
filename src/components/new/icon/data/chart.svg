<svg width="32" height="32" viewBox="0 0 32 32" xmlns="http://www.w3.org/2000/svg"><g fill-rule="evenodd"><path d="M7.4 10.54v11.132h17.2v-11.13H7.4zM24 20.473H8l.024-8.735H23.99l.01 8.735z"/><path d="M15.617 20.762l-1.542 3.69m2-3.69l1.542 3.69" stroke="currentColor" stroke-width="1.2"/><path d="M6.4 10c0 .6.494 1.1 1.1 1.1h17c.604 0 1.1-.493 1.1-1.1 0-.6-.494-1.1-1.1-1.1h-17c-.604 0-1.1.493-1.1 1.1zm4.863 8.238l.09-.106c.124-.14.23-.255.31-.33-.02.017.218-.005.036-.005.09 0 .53.004 1.592-.005h.082l1.198-.01c.43-.003.267.012.374.01l.936.694.502.365.348-.512 1.435-2.13 1.085 1.023.52.497.395-.602 2.01-3.065-1.005-.66-1.62 2.47-1.062-1.023-.512-.488-.397.584-1.48 2.198-.004.03v-.018l-.637-.462-.16-.117-.2.002-.536.004h-.127c-.42.006-.47.006-1.07.01h-.083c-1.048.01-1.718.015-1.803.015-.36 0-.525.156-1.03.74l-.1.116a18.822 18.822 0 0 0-.54.67l.95.73c.12-.155.306-.39.504-.623z"/></g></svg>
