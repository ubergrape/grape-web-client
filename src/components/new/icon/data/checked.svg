<svg width="12" height="9" viewBox="0 0 12 9" fill="none" xmlns="http://www.w3.org/2000/svg">
<path d="M4.20001 6.50199L1.69801 3.99999L0.846008 4.84599L4.20001 8.19999L11.4 0.999992L10.554 0.153992L4.20001 6.50199Z" fill="white"/>
</svg>
