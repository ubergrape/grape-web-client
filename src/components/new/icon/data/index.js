export lock from './lock.svg'
export user from './user.svg'
export userRed from './user-red.svg'
export arrowRight from './arrow-right.svg'
export search from './search.svg'
export checked from './checked.svg'
export cross from './cross.svg'

export basketball from './basketball.svg'
export bulb from './bulb.svg'
export chart from './chart.svg'
export dinner from './dinner.svg'
export education from './education.svg'
export mask from './mask.svg'
export mobile from './mobile.svg'
export news from './news.svg'
export phone from './phone.svg'
export piechart from './piechart.svg'
export prizecup from './prizecup.svg'
export rocket from './rocket.svg'
export science from './science.svg'
export sky from './sky.svg'
export smile from './smile.svg'
export tools from './tools.svg'
