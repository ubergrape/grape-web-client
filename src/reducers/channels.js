import findIndex from 'lodash/array/findIndex'
import includes from 'lodash/collection/includes'
import find from 'lodash/collection/find'
import * as types from '../constants/actionTypes'
import conf from '../conf'

const initialState = []

export default function reduce(state = initialState, action) {
  switch (action.type) {
    case types.SET_CHANNELS:
      return [
        ...state,
        ...action.payload.filter(o => !state.find(o2 => o.id === o2.id)),
      ]

    case types.SET_CHANNEL: {
      const { id, type } = action.payload.channel

      return state.reduce((newState, channel) => {
        if (channel.id === id && channel.type === type) {
          const newChannel = { ...channel, current: true }
          // In case of empty PM we're adding it to the navigation
          // with the current timestamp to sort later by it's value.
          // It is not saved in the backend and lives only
          // for the current session lifetime.
          if (type === 'pm' && !channel.firstMessageTime) {
            newChannel.temporaryInNavigation = Date.now()
          }
          newState.push(newChannel)
          return newState
        }
        if (channel.current) {
          newState.push({ ...channel, current: false })
          return newState
        }
        newState.push(channel)
        return newState
      }, [])
    }

    case types.ADD_CHANNEL: {
      const channel = action.payload
      if (find(state, { id: channel.id })) {
        return state
      }
      return [...state, channel]
    }

    case types.ADD_USER_TO_CHANNEL: {
      const { user, channelId: id, userId, isCurrentUser } = action.payload
      if (!user) return state

      const newState = [...state]
      const index = findIndex(newState, { id })
      if (index === -1) return state
      const channel = newState[index]
      const { users } = channel
      newState.splice(index, 1, {
        ...channel,
        // As a workaround of API bug,
        // we have to ensure that user isn't joined already.
        // https://github.com/ubergrape/chatgrape/issues/3804
        users: includes(users, userId) ? users : [...users, userId],
        joined: isCurrentUser ? true : channel.joined,
      })
      return newState
    }

    case types.REMOVE_USER_FROM_CHANNEL: {
      const { channelId, userId } = action.payload
      const index = findIndex(state, { id: channelId })
      if (index === -1) return state
      const channels = [...state]
      const channel = state[index]
      channels.splice(index, 1, {
        ...channel,
        users: channel.users.filter(id => id !== userId),
        joined: conf.user.id !== userId,
      })
      return channels
    }

    case types.UPDATE_CHANNEL: {
      const { id, type } = action.payload
      const newState = [...state]
      const index = findIndex(newState, { id, type })
      if (index === -1) return state
      const channel = newState[index]
      newState.splice(index, 1, {
        ...channel,
        ...action.payload,
      })
      return newState
    }

    case types.UPDATE_CHANNEL_PARTNER_INFO: {
      const { pm } = action.payload
      const newState = [...state]
      const index = findIndex(newState, { id: pm })
      if (index === -1) return state
      newState[index].partner = action.payload
      return newState
    }

    case types.REMOVE_ROOM: {
      return state.filter(
        ({ type, id }) => !(type === 'room' && id === action.payload),
      )
    }

    case types.UPDATE_CHANNEL_STATS: {
      const { channelId: id, time } = action.payload.message
      const { isCurrentUser, mentionsCount } = action.payload

      const newState = [...state]
      const index = findIndex(newState, { id })
      if (index === -1) return state
      const channel = newState[index]
      const timestamp = time.getTime()
      const mentioned = channel.mentioned || 0
      newState.splice(index, 1, {
        ...channel,
        latestMessageTime: timestamp,
        firstMessageTime: channel.firstMessageTime || timestamp,
        mentioned: mentioned + mentionsCount || channel.mentioned,
        unread: isCurrentUser ? 0 : channel.unread + 1,
      })
      return newState
    }

    case types.MARK_CHANNEL_AS_READ: {
      const { channelId: id, isCurrentUser } = action.payload
      if (!isCurrentUser) return state

      const newState = [...state]
      const index = findIndex(newState, { id })
      if (index === -1) return state
      const channel = newState[index]
      newState.splice(index, 1, {
        ...channel,
        mentioned: 0,
        unread: 0,
      })
      return newState
    }

    case types.CHANGE_FAVORITED: {
      const newState = [...state]
      action.payload.forEach(({ channelId: id, favorited }) => {
        const index = findIndex(newState, { id })
        if (index === -1) return
        const channel = newState[index]
        newState.splice(index, 1, {
          ...channel,
          favorited,
        })
      })
      return newState
    }

    case types.SET_UNSENT_MESSAGE: {
      const { id, msg } = action.payload
      const newState = [...state]
      const index = findIndex(newState, { id })
      if (index === -1) return state
      const channel = newState[index]
      newState.splice(index, 1, {
        ...channel,
        unsent: msg,
      })
      return newState
    }

    default:
      return state
  }
}
