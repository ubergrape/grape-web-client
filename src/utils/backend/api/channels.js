import rpc from '../rpc'

export const kickMemberFromChannel = (channelId, userId) =>
  rpc({
    ns: 'channels',
    action: 'kick',
    args: [channelId, userId],
  })

export const leaveChannel = channelId =>
  rpc({
    ns: 'channels',
    action: 'leave',
    args: [channelId],
  })

export const pinMessage = (channelId, messageId) =>
  rpc({
    ns: 'channels',
    action: 'pin_message',
    args: [channelId, messageId, true],
  })

export const unpinMessage = (channelId, messageId) =>
  rpc({
    ns: 'channels',
    action: 'pin_message',
    args: [channelId, messageId, false],
  })

export const loadPinnedMessages = channelId =>
  rpc(
    {
      ns: 'channels',
      action: 'get_pinned_messages',
      args: [channelId],
    },
    { camelize: true },
  )

export const getChannel = channelId =>
  rpc(
    {
      ns: 'channels',
      action: 'get',
      args: [channelId],
    },
    { camelize: true },
  )

export const deleteChannel = (channelId, name) =>
  rpc({
    ns: 'channels',
    action: 'delete',
    args: [channelId, name],
  })

export const joinChannel = channelId =>
  rpc({
    ns: 'channels',
    action: 'join',
    args: [channelId],
  })

export const inviteToChannel = (emailAddresses, channelId) =>
  rpc({
    ns: 'channels',
    action: 'invite',
    args: [channelId, emailAddresses],
  })

export const getPinnedChannels = orgId =>
  rpc(
    {
      ns: 'channels',
      action: 'get_pins',
      args: [orgId],
    },
    { camelize: true },
  )

export const addChannelToFavorites = channelId =>
  rpc({
    ns: 'channels',
    action: 'set_pin',
    args: [channelId],
  })

export const removeChannelFromFavorites = channelId =>
  rpc({
    ns: 'channels',
    action: 'remove_pin',
    args: [channelId],
  })

export const loadLatestHistory = (channelId, limit) =>
  rpc(
    {
      ns: 'channels',
      action: 'get_history',
      args: [channelId, { limit }],
    },
    { camelize: true },
  ).then(response => ({
    messages: response,
    backendHasNewerMessages: false,
  }))

export const loadOlderHistory = (channelId, limit, timeTo) =>
  rpc(
    {
      ns: 'channels',
      action: 'get_history',
      args: [channelId, { limit, timeTo }],
    },
    { camelize: true },
  ).then(response => ({
    messages: response,
    // set to undefined since this call doesn't provide any information if
    // the backend has newer messages available for this channel
    backendHasNewerMessages: undefined,
  }))

export const loadNewerHistory = (channelId, limit, timeFrom, sort) =>
  rpc(
    {
      ns: 'channels',
      action: 'get_history',
      args: [
        channelId,
        {
          // when fetching newer messages one more is fetched to identify if there are more messages
          limit: limit + 1,
          timeFrom,
          sort,
        },
      ],
    },
    { camelize: true },
  ).then(response => ({
    messages: response.slice(0, limit),
    backendHasNewerMessages: !(response.length < limit),
  }))
/**
 * Load history at a position of specified message id.
 */
export const loadHistoryAt = (channelId, messageId, options = {}) => {
  const limit = options.limit || 50
  // Amount of messages before the passed message id.
  const before = Math.round(limit / 2)
  // Amount of messages after the passed message id.
  const after = limit - before
  // Aking for one additional entry to figure if there are more messages to load
  const afterPlusOne = after + 1
  // Return an error when message id not found, otherwise return fallback results.
  const strict = true

  return rpc(
    {
      ns: 'channels',
      action: 'focus_message',
      args: [channelId, messageId, before, afterPlusOne, strict],
    },
    { camelize: true },
  ).then(response => ({
    messages: response.slice(0, limit),
    backendHasNewerMessages: !(response.length < before + afterPlusOne),
  }))
}

export const removeMessage = (channelId, messageId) =>
  rpc({
    ns: 'channels',
    action: 'delete_message',
    args: [channelId, messageId],
  })

export const updateMessage = (channelId, messageId, text) =>
  rpc({
    ns: 'channels',
    action: 'update_message',
    args: [channelId, messageId, text],
  })

export const postMessage = (channelId, text = '', options) => {
  let optionsArg = options

  if (optionsArg.attachments) {
    optionsArg = { ...optionsArg }
    // If an id is already given, like for e.g. in case of file uploads,
    // backend expect an attachment to be the id.
    // Otherwise it expects an attachment object.
    optionsArg.attachments = optionsArg.attachments.map(attachment =>
      attachment.id ? attachment.id : attachment,
    )
  }

  return rpc({
    ns: 'channels',
    action: 'post',
    args: [channelId, text, optionsArg],
  })
}

export const readMessage = (channelId, messageId) =>
  rpc({
    ns: 'channels',
    action: 'read',
    args: [channelId, messageId],
  })

export const removeLinkAttachments = (channelId, messageId, sourceUrl, type) =>
  rpc({
    ns: 'channels',
    action: 'delete_attachment',
    args: [channelId, messageId, sourceUrl, type],
  })

export const setTyping = (channelId, isTyping) =>
  rpc({
    ns: 'channels',
    action: 'set_typing',
    args: [channelId, isTyping],
  })
