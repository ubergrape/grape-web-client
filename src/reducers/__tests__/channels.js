import channels from '../channels'
import {
  SET_CHANNELS,
  UPDATE_CHANNEL_UNREAD_COUNTER,
} from '../../constants/actionTypes'

import { c1, c2, c3 } from './data/channels'

describe('channels reducer', () => {
  it('should handle SET_CHANNELS', () => {
    expect(
      channels([], {
        type: SET_CHANNELS,
        payload: [c1, c2],
      }),
    ).toMatchSnapshot()
  })

  it('should handle correct merge channels on SET_CHANNELS', () => {
    expect(
      channels([c1, c2], {
        type: SET_CHANNELS,
        payload: [c1, c3],
      }),
    ).toMatchSnapshot()
  })

  it('should handle correct update unread field on UPDATE_CHANNEL_UNREAD_COUNTER', () => {
    expect(
      channels([{ unread: 1, latestMessageTime: 1508772967000 }], {
        type: UPDATE_CHANNEL_UNREAD_COUNTER,
        payload: {
          id: 3339,
          unread: 0,
          time: 'Mon Oct 23 2017 17:36:08 GMT+0200',
        },
      }),
    ).toMatchSnapshot()
  })

  it('should return same state if channel not found on UPDATE_CHANNEL_UNREAD_COUNTER', () => {
    expect(
      channels([{ unread: 1, id: 3338, latestMessageTime: 1508772967000 }], {
        type: UPDATE_CHANNEL_UNREAD_COUNTER,
        payload: {
          id: 3339,
          unread: 0,
          time: 'Mon Oct 23 2017 17:36:08 GMT+0200',
        },
      }),
    ).toMatchSnapshot()
  })
})
