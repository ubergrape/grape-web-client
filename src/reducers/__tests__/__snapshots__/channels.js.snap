// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`channels reducer should handle SET_CHANNELS 1`] = `
Array [
  Object {
    "created": 1521822165,
    "creator": 2918,
    "icon": "https://staging.chatgrape.com/imgsrc/s/avatar_invited/200x200/",
    "id": 6476,
    "invites": Array [],
    "joined": true,
    "mentioned": 0,
    "name": "ajv13028@soioa.com",
    "type": "pm",
    "unread": 0,
    "users": Array [
      2918,
      13531,
    ],
  },
  Object {
    "abbr": "TT",
    "color": "#EA4C3A",
    "created": 1498568936,
    "creator": 2674,
    "description": "",
    "firstMessageTime": 1498568939879,
    "icon": "dinner",
    "id": 3339,
    "invites": Array [],
    "isManaged": false,
    "isPublic": true,
    "joined": false,
    "latestMessageTime": 1508772968000,
    "mentioned": 0,
    "name": "team testing",
    "slug": "team-testing",
    "type": "room",
    "unread": 0,
    "users": Array [
      2404,
    ],
  },
]
`;

exports[`channels reducer should handle correct merge channels on SET_CHANNELS 1`] = `
Array [
  Object {
    "abbr": "TT",
    "color": "#EA4C3A",
    "created": 1498568936,
    "creator": 2674,
    "description": "",
    "firstMessageTime": 1498568939879,
    "icon": "dinner",
    "id": 3339,
    "invites": Array [],
    "isManaged": false,
    "isPublic": true,
    "joined": false,
    "latestMessageTime": 1508772968000,
    "mentioned": 0,
    "name": "team testing",
    "slug": "team-testing",
    "type": "room",
    "unread": 0,
    "users": Array [
      2404,
    ],
  },
  Object {
    "created": 1521822165,
    "creator": 2918,
    "icon": "https://staging.chatgrape.com/imgsrc/s/avatar_invited/200x200/",
    "id": 6476,
    "invites": Array [],
    "joined": true,
    "mentioned": 0,
    "name": "ajv13028@soioa.com",
    "type": "pm",
    "unread": 0,
    "users": Array [
      2918,
      13531,
    ],
  },
  Object {
    "created": 1524044323,
    "creator": 2918,
    "icon": "https://staging.chatgrape.com/imgsrc/m/cebabcff380f42ea8fd2395d86dbc4a0/200x200/",
    "id": 6952,
    "invites": Array [],
    "joined": true,
    "mentioned": 0,
    "name": "asdf",
    "type": "pm",
    "unread": 0,
    "users": Array [
      2918,
      13258,
    ],
  },
]
`;

exports[`channels reducer should handle correct update unread field on UPDATE_CHANNEL_UNREAD_COUNTER 1`] = `
Array [
  Object {
    "latestMessageTime": 1508772967000,
    "unread": 1,
  },
]
`;

exports[`channels reducer should return same state if channel not found on UPDATE_CHANNEL_UNREAD_COUNTER 1`] = `
Array [
  Object {
    "id": 3338,
    "latestMessageTime": 1508772967000,
    "unread": 1,
  },
]
`;
